src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    public class Http1ContentLengthMessageBody : Http1MessageBody
+    {
+        private readonly long _contentLength;
+        private long _inputLength;
+        private ReadResult _readResult;
+        private bool _completed;
+        private int _userCanceled;
+
+        public Http1ContentLengthMessageBody(bool keepAlive, long contentLength, Http1Connection context)
+            : base(context)
+        {
+            RequestKeepAlive = keepAlive;
+            _contentLength = contentLength;
+            _inputLength = _contentLength;
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            if (_inputLength == 0)
+            {
+                _readResult = new ReadResult(default, isCanceled: false, isCompleted: true);
+                return _readResult;
+            }
+
+            TryStart();
+
+            // The while(true) loop is required because the Http1 connection calls CancelPendingRead to unblock
+            // the call to StartTimingReadAsync to check if the request timed out.
+            // However, if the user called CancelPendingRead, we want that to return a canceled ReadResult
+            // We internally track an int for that.
+            while (true)
+            {
+                // This isn't great. The issue is that TryRead can get a canceled read result
+                // which is unknown to StartTimingReadAsync. 
```

Comment:
I don't think I understand this. Shouldn't TryRead check IsCanceled and throw a BadHttpRequestException too?

Reply:
Try* API shouldn’t throw. Makes them unusable, so we should avoid it where possible

Reply:
Yeah I was trying to avoid throwing in Try methods. However, TryRead throws internally in pipes, so we should consider fixing that in pipes.

Reply:
> Try* API shouldn’t throw. Makes them unusable, so we should avoid it where possible

Like @jkotalik points out, TryRead is a little special. Errors also need to be raised from TryRead when the writer completes with an error. TryRead should only return false when the PipeReader is still active but there's no data available to currently read.

Reply:
A little more info about the TryXXX pattern: TryXXX doesn't mean the method should never throw an exception. It means it shouldn't throw an exception for a certain use-case.

For example, int.TryParse will return false if the string cannot be parsed to an integer. But it will throw an ArgumentNullException if you give it a null value.

Reply:
> TryRead should only return false when the PipeReader is still active but there's no data available to currently read.

Sure, it has to throw *sometimes* in those rare cases where the input is invalid. I don't know that I agree with the above.